design/controlling-agent-updates.md

```diff
@@ -0,0 +1,85 @@
+# Controlling agent updates
+
+The agent is considered an extension of the service.
+As such, we have an auto-update mechanism to ensure that new features can rely on the agent support they require.
+Conceptually, it's no different from binaries on the app tier or JavaScript served to the web frontend.
+
+For instance, tasks express a minimum agent version required.
+If a selected agent doesn't meet that requirement, the agent is first asked to update itself.
+Other ways an agent can be updated include using a feature which demands an evergreen agent (YAML pipelines, for instance) or clicking the "upgrade all agents" button in a pool.
+
+For some highly-regulated customers, this kind of evergreen infrastructure is at odds with their security/compliance needs.
+- For instance, Customer "K" runs the agent inside a highly change-controlled datacenter.
+Every alteration must be approved by a high-ranking official in their organization after manual source inspection.
+Today, they diff between the release tags of their current version and the version being installed, looking for potential security issues or malicious changes.
+- Customer "N" isn't as concerned about the security aspect but is in a regulated space.
+They must satisfy their auditors that they're aware of all changes taking place inside their datacenter.
+
+Such customers today are doing things like ACLing the agent's installation directory so it can't update itself or recompiling the agent without update support.
+This leaves us in a situation where no one is happy:
+we can't service the product correctly, and the customer has to maintain custom configuration / infrastructure to meet their compliance needs.
+
+## Solution
+
+_See below for other solutions considered._
+
+We will add a mode to the organization-level pool which puts it into "no update" mode.
+In this mode, jobs will use the minimum agent version as a demand.
+If no agent can satisfy the demand, the job will fail instead of forcing an agent update.
+
+The error message must clearly indicate what feature or task demanded agent update.
+It must also indicate that the customer opted into this behavior.
+This way, the customer can decide whether to update their agents or edit the pipeline to remove the offending feature.
+For example:
+> This pool doesn't allow agent updates.
+> An administrator can alter this setting on the organization pool page.
```

Comment:
If this is implemented it would be nice as a pool owner to add a reason for the lock:
```
"Agent upgrades are currently disabled due to the policy imposed by [...], If you would like to read more see <this wiki> or escalate a request to <some email alias>"
```

Personally I would love to just allow all pools to auto upgrade, but in the event of an agent bug, force a rollback to a LKG version and leave a message like:
```
"Agent versions have been pinned to 123.4 due to a bug in 123.5 pending resolution of <this icm>"
```

Right now one of the most difficult aspects of large scale pool management is the in ability to roll back or select a specific version.  If we want upgrade to the latest we can do some basic sanity checks by only upgrading 1/2 agents, but that's not enough to verify all customer scenarios.

If our agents are on 123, a customer wants 124, it would be nice to upgrade only 124, instead of jumping all the way to the latest.  Similarly, if we do upgrade to the latest and there turns out to be issues, we have to rebuild the agent. I'm not aware of a reliable means of downgrading.

Reply:
Regarding the last section about updating to intermediate versions, because rollback is not a action agents can currently take, bugfixes (and effective rollbacks) are always accompanied by a version number increase.

Reply:
I do like the idea about having the admin offer up a reason, though! Adding to the spec.